tiatoolbox/models/engine/engine_abc.py

```diff
@@ -297,119 +510,250 @@ def pre_process_patches(
+            raw_predictions[key] = raw_output
+        else:
+            raw_predictions[key] = np.append(raw_predictions[key], raw_output, axis=0)
+
+        return raw_predictions
+
     def infer_patches(
         self: EngineABC,
-        data_loader: DataLoader,
-    ) -> dict:
-        """Model inference on an image patch."""
+        dataloader: DataLoader,
+        save_path: Path | None,
+    ) -> dict | Path:
+        """Runs model inference on image patches and returns output as a dictionary.
+
+        Args:
+            dataloader (DataLoader):
+                An :class:`torch.utils.data.DataLoader` object to run inference.
+            save_path (Path | None):
+                If `cache_mode` is True then path to save zarr file must be provided.
+
+        Returns:
+            dict or Path:
+                Result of model inference as a dictionary. Returns path to
+                saved zarr file if `cache_mode` is True.
+
+        """
         progress_bar = None
 
         if self.verbose:
             progress_bar = tqdm.tqdm(
-                total=int(len(data_loader)),
+                total=int(len(dataloader)),
                 leave=True,
                 ncols=80,
                 ascii=True,
                 position=0,
             )
-        raw_predictions = {
-            "predictions": [],
-        }
+
+        keys = ["predictions"]
 
         if self.return_labels:
-            raw_predictions["labels"] = []
+            keys.append("labels")
+
+        raw_predictions = {key: None for key in keys}
+
+        zarr_group = None
+
+        if self.cache_mode:
+            zarr_group = zarr.open(save_path, mode="w")
 
-        for _, batch_data in enumerate(data_loader):
+        for _, batch_data in enumerate(dataloader):
             batch_output_predictions = self.model.infer_batch(
                 self.model,
                 batch_data["image"],
                 device=self.device,
             )
 
-            raw_predictions["predictions"].extend(batch_output_predictions.tolist())
+            raw_predictions = self._update_model_output(
+                raw_predictions=raw_predictions,
+                key="predictions",
+                raw_output=batch_output_predictions,
+            )
 
             if self.return_labels:  # be careful of `s`
-                # We do not use tolist here because label may be of mixed types
-                # and hence collated as list by torch
-                raw_predictions["labels"].extend(list(batch_data["label"]))
+                raw_predictions = self._update_model_output(
+                    raw_predictions=raw_predictions,
+                    key="labels",
+                    raw_output=batch_data["label"].numpy(),
+                )
+            if self.cache_mode:
+                zarr_group = write_to_zarr_in_cache_mode(
+                    zarr_group=zarr_group, output_data_to_save=raw_predictions
+                )
```

Comment:
raw_predictions should be set to None after updating zarr.

Reply:
Addressed in https://github.com/TissueImageAnalytics/tiatoolbox/pull/737/commits/af641e392370f1ada351d0811cae949f4b30e284